pkg/network/usm/utils/connection_rollup.go

```diff
@@ -11,169 +11,147 @@ import (
+// | (1.1.1.1:60005 - 2.2.2.2:80)  | (1.1.1.1:60001 - 2.2.2.2:80)  |
+// | (1.1.1.1:60006 - 2.2.2.2:443) | (1.1.1.1:60004 - 2.2.2.2:443) |
+// | (2.2.2.2:80 - 1.1.1.1:70000)  | (2.2.2.2:80 - 1.1.1.1:60001)  |
+// | (2.2.2.2:90001 - 1.1.1.1:80)  | (2.2.2.2:90001 - 1.1.1.1:80)  |
+// | (2.2.2.2:90002 - 1.1.1.1:80)  | (2.2.2.2:90001 - 1.1.1.1:80)  |
+// |-------------------------------+-------------------------------|
 //
-// The approach is very simplistic and only aims to address the common cases we
-// see in most workloads. The function will likely generate "false negatives",
-// but ideally it shouldn't generate "false positives". In other words, we may
-// not always rollup two different `types.ConnectionKey` that may legitimately
-// refer to the same (client, server) pair, but we shouldn't cause correctness
-// issues by aggregating two datapoints that *don't* belong to to the same
-// (client, server) pair.
+// Note that the ephemeral port numbers just happen to be the number of the
+// first "sample" processed (eg. 60001, 60004, 90001) for a given
+// (IP-A,IP-B,SERVER_PORT) tuple.
 //
-// For the sake of documentation these are the edge-cases we *don't* intend to
-// address at this time:
+// It may be convenient for the client of this code to clear those ephemera
+// numbers and have something like (1.1.1.1:0 - 2.2.2.2:80) instead of
+// (1.1.1.1:60001 - 2.2.2.2:80) for the "rolled up" key.  For that purpose we
+// provided a second method (`ClearEphemeralPort`) that can be called once the
+// stream has been processed.
 //
-// * Bare process monitoring (this code shouldn't even be reached when bare
-// process monitoring is enabled);
-// * Gossip/p2p protocols where you see symmetrical traffic like
-// (IP-A:PORT-X,IP-B:PORT-Y) and (IP-B:PORT-X,IP-A:PORT-Y);
-// * Containers running multiple processes instances bound to different ports. A
-// real world example we have here at Datadog are redis containers running
-// multiple instances. In this case only one one of the redis instances would
-// trigger a rollup.
+// NOTE: this code should not be used in the context of bare-process monitoring.
 func (c *ConnectionAggregator) RollupKey(key types.ConnectionKey) types.ConnectionKey {
 	if c == nil {
 		return key
 	}
 
-	// order IPs and ports such that
-	// srcIP < dstIP
-	//
-	// why do we do this? we want to be able to index and lookup IPs in a
-	// deterministic way, but our kernel code only does a best-effort in
-	// ordering tuples based on the *port* number, which sometimes generates
-	// tuples that look like
+	// Here we translate the ConnectionKey to a (ipTuple, portValues) pair.
+	// Note that this representation is normalized, so
 	//
-	// A:5000 B:8080
-	// B:8080 B:9000
-	//
-	// when host(A) or host(B) don't really "abide" to the correct port ranges.
-	normalizedKey, flipped := normalizeByIP(key)
+	// splitKey(IP-A:PORT-A,IP-B:PORT-B) == splitKey(IP-B:PORT-B,IP-A:PORT-A)
+	ips, ports, flipped := splitKey(key)
```

Comment:
nit: think about extracting the logic of checking if the key is flipped to a separate function and then passing it down to the function instead of returning it

Reply:
I think it makes sense to return all 3 things together, because there isn't a use-case where you call `splitKey` and disregard the `flipped` information. in other words, it doesn't make much sense to separate things because they're all used in conjunction.